### PR TITLE
CNV-1589: Provision and manage VMs using Ansible modules

### PR DESCRIPTION
**Pull request by pkliczewski**

### CODE REVIEW

**Comment by pkliczewski** [comment]
@ousleyp please review

File: cnv/cnv_users_guide/cnv-ansible-modules.adoc
Comment (alexxa) [inline cnv/cnv_users_guide/cnv-ansible-modules.adoc]: pf/of

**Review by alexxa — CHANGES_REQUESTED** [review]

**Comment by alexxa** [comment]
@pkliczewski @aburdenthehand @ousleyp Can we merge it ASAP, please? The ansible 2.8 release was on 16th.

**Comment by pkliczewski** [comment]
@alexxa I fixed the typo and I think it is ready. @ousleyp please review

**Comment by ousleyp** [comment]
> @pkliczewski @ousleyp Can we merge it ASAP, please? The ansible 2.8 release was on 16th.

Wait, is this supposed to be for CNV 1.4 docs or CNV 2.0 (the PR is against enterprise-4.1 which suggests CNV 2.0)? If it's for CNV 2.0, it won't go out for awhile regardless.

I'm sorry for the delay - I'm waiting on some info about the official way to refer to Ansible in the docs. Most of our docs just say "Ansible" like you did, but the official spreadsheet says otherwise. There are a couple other minor things that I'll note inline. 

I'll get back to you ASAP, but please let me know which version in the meantime. Thanks!

File: cnv/cnv_users_guide/cnv-ansible-modules.adoc
Comment (ousleyp) [inline cnv/cnv_users_guide/cnv-ansible-modules.adoc]: I'm waiting to find out if "Ansible" is okay or if the shortname should be "Ansible Automation" which is how it's written in the spreadsheet of official product names/shortnames. I'm not sure if the list is out-of-date or if it's a new convention that we should start following.

**Comment by pkliczewski** [comment]
@ousleyp Yes, it is CNV 2.0 item as you can see in `CNV-1589` we want to have something similar to RHV [doc](https://access.redhat.com/documentation/en-us/red_hat_virtualization/4.3/html/administration_guide/chap-automating_rhv_configuration_using_ansible).

File: cnv/cnv_users_guide/cnv-ansible-modules.adoc
Comment (ousleyp) [inline cnv/cnv_users_guide/cnv-ansible-modules.adoc]: Overall, the writing is good and you followed our conventions well (thanks for using {ProductName}! 

This particular sentence is a bit unclear to me, though. When you say it's "an alternative way" "compared to APIs and [the] `oc` tool," I'm not sure what is being compared. If you want to say that Ansible is better than using APIs or the `oc` tool, give a reason. :) Otherwise, I don't think it's necessary to mention APIs/the `oc` tool here.

Might be better to just say something like: "Ansible allows you to integrate with other Ansible modules, \<give an example or reason why you would want to do that\>."

File: cnv/cnv_users_guide/cnv-ansible-modules.adoc
Comment (ousleyp) [inline cnv/cnv_users_guide/cnv-ansible-modules.adoc]: Generally we try to avoid "see this link." Maybe the previous sentence should read:

 "...allows you to integrate link:https://docs.ansible.com/ansible/latest/modules/list_of_cloud_modules.html#kubevirt[KubeVirt modules] with other Ansible modules." 

or something like that?

File: cnv/cnv_users_guide/cnv-ansible-modules.adoc
Comment (ousleyp) [inline cnv/cnv_users_guide/cnv-ansible-modules.adoc]:
```suggestion
// Module included in the following assemblies:
//
// *

[id="cnv-ansible-modules_{context}"]
```

Comment (ousleyp) [reply cnv/cnv_users_guide/cnv-ansible-modules.adoc]: (the _{context} part is used in modules, but it's not included in assemblies.)

File: cnv/cnv_users_guide/cnv-ansible-modules.adoc
Comment (ousleyp) [inline cnv/cnv_users_guide/cnv-ansible-modules.adoc]:
```suggestion
= Automating {ProductName} management tasks by using Red Hat Ansible Automation
```

File: cnv/cnv_users_guide/cnv-ansible-modules.adoc
Comment (ousleyp) [inline cnv/cnv_users_guide/cnv-ansible-modules.adoc]:
```suggestion
```

Comment (ousleyp) [reply cnv/cnv_users_guide/cnv-ansible-modules.adoc]: We use this in assemblies to build the direct link to the module. The [id="cnv-ansible-modules_{context}"] gets its context from the :context: line of the assembly that the module is in.

Currently, there's no assembly that includes this module, but I'll create one as a follow-up PR.

File: cnv/cnv_users_guide/cnv-ansible-modules.adoc
Comment (ousleyp) [inline cnv/cnv_users_guide/cnv-ansible-modules.adoc]:
```suggestion
```

Comment (ousleyp) [reply cnv/cnv_users_guide/cnv-ansible-modules.adoc]: There isn't a table of contents in the module itself, so there's no need for `toc::[]`.

File: cnv/cnv_users_guide/cnv-ansible-modules.adoc
Comment (ousleyp) [inline cnv/cnv_users_guide/cnv-ansible-modules.adoc]:
```suggestion
Ansible includes support for {ProductName}, and Ansible modules enable you to automate
```

File: cnv/cnv_users_guide/cnv-ansible-modules.adoc
Comment (ousleyp) [inline cnv/cnv_users_guide/cnv-ansible-modules.adoc]:
```suggestion
cluster management tasks such as template, persistent volume claim, and virtual machine operations.
```

**Review by ousleyp — CHANGES_REQUESTED** [review]
Overall, looks good. Thanks for trying to follow our (confusing, ever-changing) modular docs standards. :) There are some things to fix, and I need to follow up on the naming issue, but then we should be good to merge.

File: cnv/cnv_users_guide/cnv-ansible-modules.adoc
Comment (pkliczewski) [reply cnv/cnv_users_guide/cnv-ansible-modules.adoc]: In RHV [documentation](https://access.redhat.com/documentation/en-us/red_hat_virtualization/4.3/html/administration_guide/chap-automating_rhv_configuration_using_ansible) we use `Ansible`.

File: cnv/cnv_users_guide/cnv-ansible-modules.adoc
Comment (pkliczewski) [reply cnv/cnv_users_guide/cnv-ansible-modules.adoc]: The intention was not to compare but state that it is yet another way to manage our `{ProductName}` but we have a benefit to use other modules provided by Ansible.

File: cnv/cnv_users_guide/cnv-ansible-modules.adoc
Comment (pkliczewski) [reply cnv/cnv_users_guide/cnv-ansible-modules.adoc]: Sure will update.

Comment (pkliczewski) [reply cnv/cnv_users_guide/cnv-ansible-modules.adoc]: @ousleyp What do you think about:

> Among other tools Ansible provides a way to automate {ProductName} management and allows to integrate link:https://docs.ansible.com/ansible/latest/modules/list_of_cloud_modules.html#kubevirt[KubeVirt modules] with other Ansible modules.

**Comment by alexxa** [comment]
Wait. @pkliczewski, When I asked Vatsal whether ansible modules are 1.4 or 2.0 'feature', I was told 'they work for both, and were tested for both versions.'

**Comment by pkliczewski** [comment]
@alexxa We want to document it as part of 2.0. Ansible 2.8 was released on 16th of May and it was not available before.

File: cnv/cnv_users_guide/cnv-ansible-modules.adoc
Comment (ousleyp) [reply cnv/cnv_users_guide/cnv-ansible-modules.adoc]: Thanks for letting me know. I know that in the rest of the OpenShift docs, "Ansible" is used, but it's possible the "Ansible Automation" guideline is new. I'll let you know when I have confirmation.

File: cnv/cnv_users_guide/cnv-ansible-modules.adoc
Comment (ousleyp) [reply cnv/cnv_users_guide/cnv-ansible-modules.adoc]: > Ansible provides a way to automate {ProductName} management, which you can also accomplish by using the `oc` CLI tool or APIs. Ansible is unique because it allows you to integrate link:https://docs.ansible.com/ansible/latest/modules/list_of_cloud_modules.html#kubevirt[KubeVirt modules] with other Ansible modules.

What do you think of this? If it doesn't make sense, no worries.

Comment (pkliczewski) [reply cnv/cnv_users_guide/cnv-ansible-modules.adoc]: Looks good, will update

**Comment by ousleyp** [comment]
@pkliczewski LGTM (still unsure about "Ansible Automation" vs. "Ansible" as the short name, but I can fix it later if necessary)

One last thing that I'm aware of - can you please squash your commits into one commit?

@kalexand-rh @vikram-redhat is there anything else barring us from merging this? One of us will make an assembly for it as part of [CNV-1589](https://jira.coreos.com/browse/CNV-1589).

**Comment by vikram-redhat** [comment]
@ousleyp - I am _assuming_ here that QE and peer review are done. Other than that, once we squash the commits we should be ok to merge.

**Comment by pkliczewski** [comment]
@ousleyp @vikram-redhat commits squashed. Please take a look

File: cnv/cnv_users_guide/cnv-ansible-modules.adoc
Comment (bergerhoffer) [inline cnv/cnv_users_guide/cnv-ansible-modules.adoc]: If this file is a module, then it should go in the modules/ directory.

File: cnv/cnv_users_guide/cnv-ansible-modules.adoc
Comment (bergerhoffer) [inline cnv/cnv_users_guide/cnv-ansible-modules.adoc]: And I can see that this module isn't added to an assembly yet (should it be now?), but when you're ready, then it should be updated here to list which assembly the module is a part of.

File: cnv/cnv_users_guide/cnv-ansible-modules.adoc
Comment (bergerhoffer) [inline cnv/cnv_users_guide/cnv-ansible-modules.adoc]: oc should be in backticks here, \`oc\`

File: cnv/cnv_users_guide/cnv-ansible-modules.adoc
Comment (bergerhoffer) [inline cnv/cnv_users_guide/cnv-ansible-modules.adoc]: I'm not sure if how CNV does things is different, but I think once you add this module to an assembly, I'd think the assembly should include this attribtues file, and you wouldn't include it here.

File: cnv/cnv_users_guide/cnv-ansible-modules.adoc
Comment (bergerhoffer) [inline cnv/cnv_users_guide/cnv-ansible-modules.adoc]: @ousleyp just a general comment (not for this PR) - this attribute name is a little confusing, since we have "product-title" for OpenShift and if I hadn't build it, I wouldn't have known that this was the CNV product name. I'd suggest to change this attribute name to CNVProductName or something, so that it's easier to understand from the source that it's CNV's name.

Comment (ousleyp) [reply cnv/cnv_users_guide/cnv-ansible-modules.adoc]: @bergerhoffer Good call, I have wondered if this is confusing. I can bring it up in our CNV meeting this week.

File: cnv/cnv_users_guide/cnv-ansible-modules.adoc
Comment (ousleyp) [reply cnv/cnv_users_guide/cnv-ansible-modules.adoc]: @bergerhoffer One of us from CNV docs is going to do a follow-up PR to create an assembly for this module.

File: cnv/cnv_users_guide/cnv-ansible-modules.adoc
Comment (ousleyp) [reply cnv/cnv_users_guide/cnv-ansible-modules.adoc]: Ah, excellent catch! Piotr, please move the file to the modules directory. Thanks :)

**Comment by ousleyp** [comment]
Piotr, there are just a couple of changes left:

- remove the line "include::modules/cnv-document-attributes.adoc[]"
- line 14: "the oc CLI tool" -> "the \`oc` CLI tool"
- move the file into the modules folder
- squash commits into one commit again

Then we should be able to merge. Thanks!

**Comment by pkliczewski** [comment]
@ousleyp Sure, thank you for the list. I am on it

**Comment by pkliczewski** [comment]
@ousleyp @bergerhoffer @vikram-redhat please review again all comments fixed

**Comment by ousleyp** [comment]
> @ousleyp @bergerhoffer @vikram-redhat please review again all comments fixed

LGTM. Please merge this PR if you agree, @vikram-redhat and/or @kalexand-rh :) Thanks.

**Comment by pkliczewski** [comment]
@vikram-redhat @kalexand-rh please review and let me know if you want me to fix something

**Comment by kalexand-rh** [comment]
I think this looks fine. Merging.

**Comment by vikram-redhat** [comment]
@ousleyp @kalexand-rh this module is never referenced from anywhere. As part of my https://github.com/openshift/openshift-docs/pull/16426 PR to update the CNV attributes, this is marked as deleted in the 4.2 branch. Does this seem correct?

**Comment by ousleyp** [comment]
@vikram-redhat yes, that does seem correct - I renamed the file to modules/cnv-about-red-hat-ansible-automation.adoc in https://github.com/openshift/openshift-docs/pull/15956

I'll go ahead and delete modules/cnv-ansible-modules.adoc in that branch again - no worries.

**Pull request merged**
